src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java

```diff
@@ -136,11 +152,29 @@ public void afterReload() {
     }
 
     private void resetCaches(final Iterable<Object> configValues) {
+        forEachValue(configValues, configValue -> {
+            // Only clear the caches of configs that don't need a restart
+            if (configValue.getSpec().restartType == RestartType.NONE) {
+                configValue.clearCache();
+            }
+        });
+    }
+
+    private void forEachValue(Iterable<Object> configValues, Consumer<ConfigValue<?>> consumer) {
         configValues.forEach(value -> {
             if (value instanceof ConfigValue<?> configValue) {
-                configValue.clearCache();
+                consumer.accept(configValue);
             } else if (value instanceof Config innerConfig) {
-                this.resetCaches(innerConfig.valueMap().values());
+                forEachValue(innerConfig.valueMap().values(), consumer);
+            }
+        });
+    }
+
+    @ApiStatus.Internal
+    public void resetWorldCaches() {
+        forEachValue(getValues().valueMap().values(), configValue -> {
+            if (configValue.getSpec().restartType == RestartType.WORLD) {
+                configValue.clearCache();
             }
         });
     }
```

Comment:
This whole thing can be implemented using a single `resetCaches(Iterable<Object> configValues, RestartType restartType)` function.

Reply:
`forEachValue` is used by the validator, the best I can do is merge the 2 resetCache methods.